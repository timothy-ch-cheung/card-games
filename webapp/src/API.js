import axios from 'axios';

const instance = axios.create({
    baseURL: 'http://127.0.0.1:8080'
});

export default instance;