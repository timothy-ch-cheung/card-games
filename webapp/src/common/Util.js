import {act} from "@testing-library/react";

export function isBlank(str) {
    return (!str || /^\s*$/.test(str));
}

export const flushPromises = () => new Promise(resolve => setTimeout(resolve))

function wait(amount = 0) {
    return new Promise(resolve => setTimeout(resolve, amount));
}

export async function actWait(amount = 0) {
    await act(async () => {
        await wait(amount);
    });
}