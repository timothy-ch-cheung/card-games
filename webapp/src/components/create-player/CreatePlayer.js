import Modal from "react-bootstrap/Modal";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import * as React from "react";
import {useState} from "react";
import {useDispatch} from "react-redux";
import API from "../../API";
import {setKey, setPlayer} from "../../redux/actions";

function CreatePlayer(props) {

    const dispatch = useDispatch();
    const [validated, setValidated] = useState(false);

    const onClose = e => {
        props.onClose && props.onClose(e);
        setValidated(false);
    };

    const submitHandler = e => {
        e.preventDefault();
        const form = e.currentTarget;
        const valid = form.checkValidity();
        if (valid === false) {
            e.stopPropagation();
            setValidated(true);
            return;
        }
        setValidated(false);
        props.onClose();
        API.post('/player', {
            username: e.target.nickname.value
        }).then(function (response) {
            console.log(response)
            dispatch(setPlayer(response.data.id));
            dispatch(setKey(response.data.key));
            props.onSubmit(props.gameId, response.data.id, response.data.key);
        }).catch(function (error) {
            console.log(error);
        });
    };

    return (
        <div>
            <style type="text/css">
                {`
                        .modal-header button {
                           display: none;
                        }
                        
                        .btn-xxl {
                            font-size: 1.6rem;
                            border: grey solid 1px;
                            width: 220px;
                            margin: 20px 10px 30px;
                        }
                        
                        .left-title {
                            text-align: left;
                        }
                        
                        @media screen and (max-width: 768px) {
                            .btn-xxl {
                                font-size: 1.3rem;
                                width: 170px;
                            }
                        }
                        
                        .create-player-modal {
                            margin-top: 50px;
                        }
                        
                        input[type=text] {
                            margin-bottom: 5px;
                        }
                    `}
            </style>
            <Modal className="create-player-modal" show={props.show} onHide={props.onClose} data-test="create-player-modal">
                <Modal.Header closeButton>
                    <Modal.Title>Set Nickname</Modal.Title>
                </Modal.Header>
                <Form noValidate validated={validated} onSubmit={submitHandler}>
                    <Modal.Body>
                        <Form.Label>Nickname</Form.Label>
                        <Form.Control required type="text" placeholder="name" name="nickname"/>
                    </Modal.Body>

                    <Modal.Footer>
                        <Button variant="secondary" onClick={onClose} data-test="close-create-player-modal">Close</Button>
                        <Button type="submit" variant="info" data-test="submit-create-player-modal">Join Game</Button>
                    </Modal.Footer>
                </Form>
            </Modal>
        </div>
    )
}

export default CreatePlayer;