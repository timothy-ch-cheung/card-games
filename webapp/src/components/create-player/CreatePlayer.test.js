import React from 'react';
import configureStore from "redux-mock-store";
import {mount} from "enzyme";
import {Provider} from "react-redux";
import CreatePlayer from "./CreatePlayer";
import API from "../../API";
import {flushPromises} from "../../common/Util";

describe("TEST SUITE Create Player: ", () => {
    const mockStore = configureStore([]);
    let store;
    let wrapper;
    const MockAdapter = require("axios-mock-adapter");
    let mockAPI;

    let close;
    let submit;

    beforeAll(() => {
        mockAPI = new MockAdapter(API);
    });

    beforeEach(() => {
        close = jest.fn();
        submit = jest.fn();
        store = mockStore({gameMode: "MATCH_TWO", game: 1});
        wrapper = mount(<Provider store={store}><CreatePlayer onClose={close} show={true}
                                                              onSubmit={submit}/></Provider>);
        mockAPI.onPost("/player").reply(200, {
            id: "12345678901234567890123456789023",
            key: "keykeykeykeykeykeykeykeykeykeyke"
        });
    });

    afterEach(() => {
        store.clearActions();
        mockAPI.resetHistory();
        mockAPI.reset();
    });

    test("Matches snapshot", () => {
        expect(wrapper).toMatchSnapshot()
    });

    test("Calls onClose when close button is clicked", () => {
        wrapper.find('button[data-test="close-create-player-modal"]').simulate('click');
        expect(close).toHaveBeenCalledTimes(1);
    });

    test("Should make POST request and save response when from submitted", async () => {
        const eventObj = {
            preventDefault: jest.fn(),
            target: {
                nickname: {value: 'John'}
            }
        };
        let form = wrapper.find('form');
        form.getDOMNode().checkValidity = jest.fn(() => true);
        form.simulate('submit', eventObj);
        await flushPromises();

        expect(submit).toHaveBeenCalledTimes(1);
        expect(mockAPI.history.post.length).toBe(1);
        let actions = store.getActions();
        expect(actions.length).toBe(2);
        expect(actions[0]).toEqual({type: "SET_PLAYER", payload: "12345678901234567890123456789023"});
        expect(actions[1]).toEqual({type: "SET_KEY", payload: "keykeykeykeykeykeykeykeykeykeyke"});
    });

    test("Should not make POST request when form validation", async () => {
        let form = wrapper.find('form');
        form.getDOMNode().checkValidity = jest.fn(() => false);
        form.simulate('submit', {preventDefault: jest.fn()});
        await flushPromises();

        expect(submit).toHaveBeenCalledTimes(0);
        expect(mockAPI.history.post.length).toBe(0);
        expect(store.getActions().length).toBe(0);
    });
});