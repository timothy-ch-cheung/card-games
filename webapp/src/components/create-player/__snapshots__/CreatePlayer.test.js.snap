// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST SUITE Create Player:  Matches snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CreatePlayer
    onClose={[MockFunction]}
    onSubmit={[MockFunction]}
    show={true}
  >
    <div>
      <style
        type="text/css"
      >
        
                        .modal-header button {
                           display: none;
                        }
                        
                        .btn-xxl {
                            font-size: 1.6rem;
                            border: grey solid 1px;
                            width: 220px;
                            margin: 20px 10px 30px;
                        }
                        
                        .left-title {
                            text-align: left;
                        }
                        
                        @media screen and (max-width: 768px) {
                            .btn-xxl {
                                font-size: 1.3rem;
                                width: 170px;
                            }
                        }
                        
                        .create-player-modal {
                            margin-top: 50px;
                        }
                        
                        input[type=text] {
                            margin-bottom: 5px;
                        }
                    
      </style>
      <Bootstrap(Modal)
        className="create-player-modal"
        data-test="create-player-modal"
        onHide={[MockFunction]}
        show={true}
      >
        <Modal
          animation={true}
          autoFocus={true}
          backdrop={true}
          bsPrefix="modal"
          className="create-player-modal"
          data-test="create-player-modal"
          dialogAs={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "ModalDialog",
              "render": [Function],
            }
          }
          enforceFocus={true}
          keyboard={true}
          onHide={[MockFunction]}
          restoreFocus={true}
          show={true}
        >
          <Modal
            autoFocus={true}
            backdrop={true}
            backdropTransition={[Function]}
            className="create-player-modal modal"
            containerClassName="modal-open"
            enforceFocus={true}
            keyboard={true}
            manager={
              BootstrapModalManager {
                "adjustAndStore": [Function],
                "containers": Array [
                  <body
                    class="modal-open"
                    style="padding-right: 0px;"
                  >
                    <div
                      class="fade modal-backdrop show"
                    />
                    <div
                      aria-modal="true"
                      class="fade create-player-modal modal show"
                      role="dialog"
                      style="display: block;"
                      tabindex="-1"
                    >
                      <div
                        class="modal-dialog"
                        data-test="create-player-modal"
                        role="document"
                      >
                        <div
                          class="modal-content"
                        >
                          <div
                            class="modal-header"
                          >
                            <div
                              class="modal-title h4"
                            >
                              Set Nickname
                            </div>
                            <button
                              class="close"
                              type="button"
                            >
                              <span
                                aria-hidden="true"
                              >
                                ×
                              </span>
                              <span
                                class="sr-only"
                              >
                                Close
                              </span>
                            </button>
                          </div>
                          <form
                            class=""
                            novalidate=""
                          >
                            <div
                              class="modal-body"
                            >
                              <label
                                class="form-label"
                              >
                                Nickname
                              </label>
                              <input
                                class="form-control"
                                name="nickname"
                                placeholder="name"
                                required=""
                                type="text"
                              />
                            </div>
                            <div
                              class="modal-footer"
                            >
                              <button
                                class="btn btn-secondary"
                                data-test="close-create-player-modal"
                                type="button"
                              >
                                Close
                              </button>
                              <button
                                class="btn btn-info"
                                data-test="submit-create-player-modal"
                                type="submit"
                              >
                                Join Game
                              </button>
                            </div>
                          </form>
                        </div>
                      </div>
                    </div>
                  </body>,
                ],
                "data": Array [
                  Object {
                    "classes": Array [
                      "modal-open",
                    ],
                    "modals": Array [
                      Object {
                        "add": [Function],
                        "backdrop": <div
                          class="fade modal-backdrop show"
                        />,
                        "dialog": <div
                          aria-modal="true"
                          class="fade create-player-modal modal show"
                          role="dialog"
                          style="display: block;"
                          tabindex="-1"
                        >
                          <div
                            class="modal-dialog"
                            data-test="create-player-modal"
                            role="document"
                          >
                            <div
                              class="modal-content"
                            >
                              <div
                                class="modal-header"
                              >
                                <div
                                  class="modal-title h4"
                                >
                                  Set Nickname
                                </div>
                                <button
                                  class="close"
                                  type="button"
                                >
                                  <span
                                    aria-hidden="true"
                                  >
                                    ×
                                  </span>
                                  <span
                                    class="sr-only"
                                  >
                                    Close
                                  </span>
                                </button>
                              </div>
                              <form
                                class=""
                                novalidate=""
                              >
                                <div
                                  class="modal-body"
                                >
                                  <label
                                    class="form-label"
                                  >
                                    Nickname
                                  </label>
                                  <input
                                    class="form-control"
                                    name="nickname"
                                    placeholder="name"
                                    required=""
                                    type="text"
                                  />
                                </div>
                                <div
                                  class="modal-footer"
                                >
                                  <button
                                    class="btn btn-secondary"
                                    data-test="close-create-player-modal"
                                    type="button"
                                  >
                                    Close
                                  </button>
                                  <button
                                    class="btn btn-info"
                                    data-test="submit-create-player-modal"
                                    type="submit"
                                  >
                                    Join Game
                                  </button>
                                </div>
                              </form>
                            </div>
                          </div>
                        </div>,
                        "isTopModal": [Function],
                        "remove": [Function],
                        "setBackdropRef": [Function],
                        "setDialogRef": [Function],
                      },
                    ],
                    "overflowing": true,
                    "style": Object {
                      "overflow": "",
                      "paddingRight": "",
                    },
                  },
                ],
                "handleContainerOverflow": true,
                "hideSiblingNodes": true,
                "modals": Array [
                  Object {
                    "add": [Function],
                    "backdrop": <div
                      class="fade modal-backdrop show"
                    />,
                    "dialog": <div
                      aria-modal="true"
                      class="fade create-player-modal modal show"
                      role="dialog"
                      style="display: block;"
                      tabindex="-1"
                    >
                      <div
                        class="modal-dialog"
                        data-test="create-player-modal"
                        role="document"
                      >
                        <div
                          class="modal-content"
                        >
                          <div
                            class="modal-header"
                          >
                            <div
                              class="modal-title h4"
                            >
                              Set Nickname
                            </div>
                            <button
                              class="close"
                              type="button"
                            >
                              <span
                                aria-hidden="true"
                              >
                                ×
                              </span>
                              <span
                                class="sr-only"
                              >
                                Close
                              </span>
                            </button>
                          </div>
                          <form
                            class=""
                            novalidate=""
                          >
                            <div
                              class="modal-body"
                            >
                              <label
                                class="form-label"
                              >
                                Nickname
                              </label>
                              <input
                                class="form-control"
                                name="nickname"
                                placeholder="name"
                                required=""
                                type="text"
                              />
                            </div>
                            <div
                              class="modal-footer"
                            >
                              <button
                                class="btn btn-secondary"
                                data-test="close-create-player-modal"
                                type="button"
                              >
                                Close
                              </button>
                              <button
                                class="btn btn-info"
                                data-test="submit-create-player-modal"
                                type="submit"
                              >
                                Join Game
                              </button>
                            </div>
                          </form>
                        </div>
                      </div>
                    </div>,
                    "isTopModal": [Function],
                    "remove": [Function],
                    "setBackdropRef": [Function],
                    "setDialogRef": [Function],
                  },
                ],
                "restore": [Function],
                "scrollbarSize": 0,
              }
            }
            onClick={[Function]}
            onEnter={[Function]}
            onEntering={[Function]}
            onExited={[Function]}
            onHide={[MockFunction]}
            onMouseUp={[Function]}
            renderBackdrop={[Function]}
            restoreFocus={true}
            show={true}
            style={
              Object {
                "paddingLeft": undefined,
                "paddingRight": undefined,
              }
            }
            transition={[Function]}
          >
            <Portal
              containerInfo={
                <body
                  class="modal-open"
                  style="padding-right: 0px;"
                >
                  <div
                    class="fade modal-backdrop show"
                  />
                  <div
                    aria-modal="true"
                    class="fade create-player-modal modal show"
                    role="dialog"
                    style="display: block;"
                    tabindex="-1"
                  >
                    <div
                      class="modal-dialog"
                      data-test="create-player-modal"
                      role="document"
                    >
                      <div
                        class="modal-content"
                      >
                        <div
                          class="modal-header"
                        >
                          <div
                            class="modal-title h4"
                          >
                            Set Nickname
                          </div>
                          <button
                            class="close"
                            type="button"
                          >
                            <span
                              aria-hidden="true"
                            >
                              ×
                            </span>
                            <span
                              class="sr-only"
                            >
                              Close
                            </span>
                          </button>
                        </div>
                        <form
                          class=""
                          novalidate=""
                        >
                          <div
                            class="modal-body"
                          >
                            <label
                              class="form-label"
                            >
                              Nickname
                            </label>
                            <input
                              class="form-control"
                              name="nickname"
                              placeholder="name"
                              required=""
                              type="text"
                            />
                          </div>
                          <div
                            class="modal-footer"
                          >
                            <button
                              class="btn btn-secondary"
                              data-test="close-create-player-modal"
                              type="button"
                            >
                              Close
                            </button>
                            <button
                              class="btn btn-info"
                              data-test="submit-create-player-modal"
                              type="submit"
                            >
                              Join Game
                            </button>
                          </div>
                        </form>
                      </div>
                    </div>
                  </div>
                </body>
              }
            >
              <BackdropTransition
                appear={true}
                in={true}
              >
                <Fade
                  appear={true}
                  in={true}
                  mountOnEnter={false}
                  timeout={300}
                  unmountOnExit={false}
                >
                  <Transition
                    addEndListener={[Function]}
                    appear={true}
                    enter={true}
                    exit={true}
                    in={true}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    timeout={300}
                    unmountOnExit={false}
                  >
                    <div
                      className="fade modal-backdrop show"
                      onClick={[Function]}
                    />
                  </Transition>
                </Fade>
              </BackdropTransition>
              <DialogTransition
                appear={true}
                in={true}
                onEnter={[Function]}
                onEntering={[Function]}
                onExited={[Function]}
                unmountOnExit={true}
              >
                <Fade
                  appear={true}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntering={[Function]}
                  onExited={[Function]}
                  timeout={300}
                  unmountOnExit={true}
                >
                  <Transition
                    addEndListener={[Function]}
                    appear={true}
                    enter={true}
                    exit={true}
                    in={true}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    timeout={300}
                    unmountOnExit={true}
                  >
                    <div
                      aria-modal={true}
                      className="fade create-player-modal modal show"
                      onClick={[Function]}
                      onMouseUp={[Function]}
                      role="dialog"
                      style={
                        Object {
                          "paddingLeft": undefined,
                          "paddingRight": undefined,
                        }
                      }
                      tabIndex={-1}
                    >
                      <ModalDialog
                        data-test="create-player-modal"
                        onMouseDown={[Function]}
                        role="document"
                      >
                        <div
                          className="modal-dialog"
                          data-test="create-player-modal"
                          onMouseDown={[Function]}
                          role="document"
                        >
                          <div
                            className="modal-content"
                          >
                            <ModalHeader
                              closeButton={true}
                              closeLabel="Close"
                            >
                              <div
                                className="modal-header"
                              >
                                <ModalTitle>
                                  <ForwardRef
                                    className="modal-title"
                                  >
                                    <div
                                      className="modal-title h4"
                                    >
                                      Set Nickname
                                    </div>
                                  </ForwardRef>
                                </ModalTitle>
                                <CloseButton
                                  label="Close"
                                  onClick={[Function]}
                                >
                                  <button
                                    className="close"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    <span
                                      aria-hidden="true"
                                    >
                                      ×
                                    </span>
                                    <span
                                      className="sr-only"
                                    >
                                      Close
                                    </span>
                                  </button>
                                </CloseButton>
                              </div>
                            </ModalHeader>
                            <Form
                              inline={false}
                              noValidate={true}
                              onSubmit={[Function]}
                              validated={false}
                            >
                              <form
                                className=""
                                noValidate={true}
                                onSubmit={[Function]}
                              >
                                <ModalBody>
                                  <div
                                    className="modal-body"
                                  >
                                    <FormLabel
                                      column={false}
                                      srOnly={false}
                                    >
                                      <label
                                        className="form-label"
                                      >
                                        Nickname
                                      </label>
                                    </FormLabel>
                                    <FormControl
                                      name="nickname"
                                      placeholder="name"
                                      required={true}
                                      type="text"
                                    >
                                      <input
                                        className="form-control"
                                        name="nickname"
                                        placeholder="name"
                                        required={true}
                                        type="text"
                                      />
                                    </FormControl>
                                  </div>
                                </ModalBody>
                                <ModalFooter>
                                  <div
                                    className="modal-footer"
                                  >
                                    <Button
                                      active={false}
                                      data-test="close-create-player-modal"
                                      disabled={false}
                                      onClick={[Function]}
                                      type="button"
                                      variant="secondary"
                                    >
                                      <button
                                        className="btn btn-secondary"
                                        data-test="close-create-player-modal"
                                        disabled={false}
                                        onClick={[Function]}
                                        type="button"
                                      >
                                        Close
                                      </button>
                                    </Button>
                                    <Button
                                      active={false}
                                      data-test="submit-create-player-modal"
                                      disabled={false}
                                      type="submit"
                                      variant="info"
                                    >
                                      <button
                                        className="btn btn-info"
                                        data-test="submit-create-player-modal"
                                        disabled={false}
                                        type="submit"
                                      >
                                        Join Game
                                      </button>
                                    </Button>
                                  </div>
                                </ModalFooter>
                              </form>
                            </Form>
                          </div>
                        </div>
                      </ModalDialog>
                    </div>
                  </Transition>
                </Fade>
              </DialogTransition>
            </Portal>
          </Modal>
        </Modal>
      </Bootstrap(Modal)>
    </div>
  </CreatePlayer>
</Provider>
`;
